### ISSUE TITLE
Suggestion: decimal.Sub It's better not to rescale if d1.exp == d2.exp

### ISSUE THREAD

**Issue by lvhuat** [shopspring/decimal#128]
OriginCode:
```
func (d Decimal) Sub(d2 Decimal) Decimal {
	baseScale := min(d.exp, d2.exp)
	rd := d.rescale(baseScale)
	rd2 := d2.rescale(baseScale)

	d3Value := new(big.Int).Sub(rd.value, rd2.value)
	return Decimal{
		value: d3Value,
		exp:   baseScale,
	}
}
```

**Comment by kempeng** [comment]
I agree, and I propose this simple, but (very) effective test being added to rescale():

```
// NOTE(vadim): must convert exps to int64 before - to prevent overflow
	diff := abs64(int64(exp) - int64(d.exp))
// NEW TEST, avoids unnecessary rescaling
	if diff == 0 {
		return d
	}
	value := new(big.Int).Set(d.value)
	expScale := new(big.Int).Exp(tenInt, big.NewInt(int64(diff)), nil)
```

**Comment by kempeng** [comment]
To provide a bit more context, most math function start with:
```
        baseScale := min(d.exp, d2.exp)
	rd := d.rescale(baseScale)
	rd2 := d2.rescale(baseScale)
```
This results in rescale of both decimals. One rescale is always unnecessary, other only if its scale differ. The proposed test in rescale() avoids both avoiding unnessecary allocation of a new big.Int in all cases, speeding most decimal math operations up by at least 50%, by avoiding memory allocation & GC.

**Comment by mwoss** [comment]
Closed by #160
